bpf/generictracer/python.c

```diff
@@ -0,0 +1,68 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//go:build obi_bpf_ignore
+
+#include <bpfcore/vmlinux.h>
+#include <bpfcore/bpf_helpers.h>
+
+#include <logger/bpf_dbg.h>
+
+#include <common/runtime.h>
+#include <common/trace_key.h>
+#include <common/tp_info.h>
+#include <pid/pid.h>
+#include <maps/python_async_context.h>
+#include <maps/server_traces.h>
+
+SEC("uprobe/libpython3.so:context_run")
+int BPF_UPROBE(obi_uprobe_context_run, void *context) {
+    (void)ctx;
+
+    u64 id = bpf_get_current_pid_tgid();
+
+    if (!valid_pid(id)) {
+        return 0;
+    }
+
+    // find the trace key for this thread
+    trace_key_t t_key = {0};
+    task_tid(&t_key.p_key);
+    t_key.extra_id = extra_runtime_id();
+
+    u64 context_ptr = (u64)context;
+
+    bpf_dbg_printk("=== uprobe Python PyContext_Run id=%d context=%llx ===", id, context);
+    bpf_map_update_elem(&current_python_context_map, &t_key, &context_ptr, BPF_ANY);
+    return 0;
+}
+
+SEC("uprobe/libpython3.so:copy_current_ret")
+int BPF_URETPROBE(obi_uprobe_copy_current_ret, void *context) {
```

Comment:
We can't use those, we must always use UPROBE because all of these dynamic languages can grow the stack. Since uprobes bind to the return address of the function, if the stack grows it will crash the application. The current "uretprobes" are actually uprobes, where we search for the "ret" opcodes in the function and attach a regular "uprobe" to those. If you couldn't attach uprobes to the return of copy_current_ret, then it must do a jump an exit through another function. We need to disassemble source of the python binary and look at what happens.